pkg/minikube/sshutil/sshutil.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sshutil
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/docker/machine/libmachine/drivers"
+	machinessh "github.com/docker/machine/libmachine/ssh"
+	"golang.org/x/crypto/ssh"
+)
+
+// SSHSession provides methods for running commands on a host.
+type SSHSession interface {
+	Close() error
+	StdinPipe() (io.WriteCloser, error)
+	Run(cmd string) error
+	Wait() error
+}
+
+// NewSSHClient returns an SSH client object for running commands.
+func NewSSHClient(d drivers.Driver) (*ssh.Client, error) {
+	h, err := newSSHHost(d)
+	if err != nil {
+		return nil, err
+
+	}
+	auth := &machinessh.Auth{}
+	if h.SSHKeyPath != "" {
+		auth.Keys = []string{h.SSHKeyPath}
+	}
+	config, err := machinessh.NewNativeConfig(h.Username, auth)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := ssh.Dial("tcp", fmt.Sprintf("%s:%d", h.IP, h.Port), &config)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// Transfer uses an SSH session to copy a file to the remote machine.
+func Transfer(data []byte, remotedir, filename string, perm string, c *ssh.Client) error {
+	// Delete the old file first. This makes sure permissions get reset.
+	deleteCmd := fmt.Sprintf("sudo rm -f %s", filepath.Join(remotedir, filename))
+	if err := RunCommand(c, deleteCmd); err != nil {
+		return err
+	}
+
+	s, err := c.NewSession()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		w, err := s.StdinPipe()
+		defer w.Close()
+		if err != nil {
+			return
+		}
+		header := fmt.Sprintf("C%s %d %s\n", perm, len(data), filename)
+		fmt.Fprint(w, header)
+		reader := bytes.NewReader(data)
+		io.Copy(w, reader)
+		fmt.Fprint(w, "\x00")
+	}()
+
+	scpcmd := fmt.Sprintf("sudo /usr/local/bin/scp -t %s", remotedir)
```

Comment:
What about if `scp` is located in another location?
Maybe use a shell or query `which scp` first.
(This isn't so important...)